yt_dlp/extractor/facebook.py

```diff
@@ -1017,19 +1034,31 @@ def _real_extract(self, url):
         video_id = self._match_id(url)
         webpage = self._download_webpage(url, video_id)
 
-        post_data = [self._parse_json(j, video_id, fatal=False)
-                     for j in re.findall(r's\.handle\(({.*})\);requireLazy\(', webpage)]
-        data = traverse_obj(post_data, (
-            ..., 'require', ..., ..., ..., 'props', 'deeplinkAdCard', 'snapshot', {dict}), get_all=False)
+        if post_data := [self._parse_json(j, video_id, fatal=False)
+                         for j in re.findall(r'data-sjs>({.*?ScheduledServerJS.*?})</script>', webpage)]:
+            data = traverse_obj(post_data, (
+                ..., 'require', ..., ..., ..., '__bbox', 'require', ..., ..., ...,
+                'entryPointRoot', 'otherProps', 'deeplinkAdCard', 'snapshot', {dict}), get_all=False)
+        elif post_data := [self._parse_json(j, video_id, fatal=False)
+                           for j in re.findall(r's\.handle\(({.*})\);requireLazy\(', webpage)]:
+            data = traverse_obj(post_data, (
+                ..., 'require', ..., ..., ..., 'props', 'deeplinkAdCard', 'snapshot', {dict}), get_all=False)
         if not data:
             raise ExtractorError('Unable to extract ad data')
 
         title = data.get('title')
         if not title or title == '{{product.name}}':
             title = join_nonempty('display_format', 'page_name', delim=' by ', from_dict=data)
+        markup = None
+        if markup_id := traverse_obj(data, ('body', '__m', {str_or_none}), get_all=False):
+            markup = clean_html(traverse_obj(post_data, (
+                ..., 'require', ..., ..., ..., '__bbox', 'markup', lambda _, v: v[0].startswith(markup_id),
+                ..., '__html', {lambda x: x if x and not x.startswith('&#123;&#123;product.') else None}),
+                get_all=False))
```

Comment:
Let's also simplify this... get rid of the assignment expression and instead just rely on traversal catching the `.startswith(None)` error and returning `None`. (We shouldn't be using `str_or_none` in the `markup_id` traversal path anyways, expecting the type is better). We can also transform with `clean_html` inside the `markup` traversal path, which makes everything a bit easier to read:

```suggestion
        markup_id = traverse_obj(data, ('body', '__m', {str}))
        markup = traverse_obj(post_data, (
            ..., 'require', ..., ..., ..., '__bbox', 'markup', lambda _, v: v[0].startswith(markup_id),
            ..., '__html', {clean_html}, {lambda x: not x.startswith('{{product.') and x}, any))
```